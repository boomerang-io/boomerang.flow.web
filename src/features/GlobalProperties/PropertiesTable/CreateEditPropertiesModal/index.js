import React, { Component } from "react";
import PropTypes from "prop-types";
import { ModalFlow } from "@boomerang/carbon-addons-boomerang-react";
import { Button } from "carbon-components-react";
import CreateEditPropertiesContent from "./CreateEditPropertiesContent";
import { Add16 } from "@carbon/icons-react";

class CreateEditPropertiesModal extends Component {
  static propTypes = {
    addPropertyInStore: PropTypes.func.isRequired,
    isOpen: PropTypes.bool,
    isEdit: PropTypes.bool,
    property: PropTypes.object,
    properties: PropTypes.array,
    handleEditClose: PropTypes.func,
    updatePropertyInStore: PropTypes.func.isRequired
  };

  render() {
    const {
      addPropertyInStore,
      isEdit,
      isOpen,
      handleEditClose,
      property,
      properties,
      updatePropertyInStore
    } = this.props;

    /**
     * arrays of values for making the key unique
     * filter out own value if editing a property, pass through all if creating
     */
    let propertyKeys = [];
    if (Array.isArray(properties)) {
      propertyKeys = properties.map(propertyObj => propertyObj.key);
      if (isEdit && property) {
        propertyKeys = propertyKeys.filter(propertyItem => propertyItem !== property.key);
      }
    }

    return (
      <ModalFlow
        isOpen={isOpen}
        modalTrigger={({ openModal }) =>
          !isEdit ? (
            <Button
              iconDescription="Create Property"
              renderIcon={Add16}
              style={{ width: "12rem" }}
              size="field"
              onClick={openModal}
              data-testid="create-global-configurations-property-button"
            >
              Create Property
            </Button>
          ) : null
        }
        modalHeaderProps={{
          title: isEdit && property ? `Edit ${property.label}` : "Create Property"
        }}
        confirmModalProps={{
          title: "Are you sure?",
          children: "Your property will not be saved",
          affirmativeAction: isEdit ? handleEditClose : null
        }}
      >
        <CreateEditPropertiesContent
          addPropertyInStore={addPropertyInStore}
          handleEditClose={handleEditClose}
          isEdit={isEdit}
          property={property}
          propertyKeys={propertyKeys}
          updatePropertyInStore={updatePropertyInStore}
        />
      </ModalFlow>
    );
  }
}

export default CreateEditPropertiesModal;
