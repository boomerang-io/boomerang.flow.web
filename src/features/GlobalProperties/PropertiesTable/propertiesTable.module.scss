@import "~@boomerang/boomerang-styles/lib/colors";
@import "~Styles/variables";

.search {
  margin-bottom: 1.31rem;
  width: 20rem;
}
.header {
  display: flex;
  justify-content: space-between;
}

.tableContainer {
  margin: 2.5rem 2.5rem 4.5rem 2.5rem;
}

.tableHeadHeader {
  font-size: inherit;
  font-family: inherit;
  height: 3rem;

  button {
    width: 100% !important;
  }

  &.label {
    width: 20%;
  }

  &.key {
    width: 20%;
  }

  &.description {
    width: 25%;
  }

  &.value {
    width: 25%;
  }

  &.secured {
    width: 5%;
  }
}

.tableBody {
  font-weight: 300;
}

.tableCell {
  align-items: center;
  display: flex;
  height: 3.625rem;
  line-height: 1.2;
  overflow-y: auto;
  padding-left: 0.75rem;
  width: 100%;

  .tableSecured {
    height: 1.25rem;
    width: 1.25rem;
  }

  .secured {
    path {
      fill: var(--cds-support-02);
    }
  }

  .unsecured {
    path {
      fill: var(--cds-support-01);
    }
  }
}

.tableTextarea {
  background: inherit;
  margin-left: 0.25rem;
  max-width: 100%;
  resize: none;
  text-align: left;
  white-space: normal;
  width: 90%;
  word-break: break-all;
}
