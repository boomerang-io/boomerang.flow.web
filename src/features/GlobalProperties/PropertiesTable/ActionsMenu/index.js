import React, { useState } from "react";
import { OverflowMenu, OverflowMenuItem } from "carbon-components-react";
import { ConfirmModal } from "@boomerang/carbon-addons-boomerang-react";
import CreateEditPropertiesModal from "../CreateEditPropertiesModal";
import "./actionMenu.module.scss";

const OverflowMenuComponent = ({ property, properties, deleteProperty, addPropertyInStore, updatePropertyInStore }) => {
  const [deleteModalIsOpen, setDeleteModalIsOpen] = useState(false);
  const [editModalIsOpen, setEditModalIsOpen] = useState(false);

  const menuOptions = [
    {
      itemText: "Edit",
      onClick: () => setEditModalIsOpen(true),
      primaryFocus: true
    },
    {
      itemText: "Delete",
      onClick: () => setDeleteModalIsOpen(true),
      hasDivider: true,
      isDelete: true
    }
  ];

  const handleEditClose = () => {
    setEditModalIsOpen(false);
  };

  return (
    <>
      <OverflowMenu
        ariaLabel="Overflow menu"
        iconDescription="Overflow menu icon"
        data-testid="configuration-property-table-overflow-menu"
        flipped
      >
        {menuOptions.map((option, index) => (
          <OverflowMenuItem key={index} {...option} />
        ))}
      </OverflowMenu>
      {editModalIsOpen && (
        <CreateEditPropertiesModal
          isOpen
          isEdit
          property={property}
          properties={properties}
          addPropertyInStore={addPropertyInStore}
          updatePropertyInStore={updatePropertyInStore}
          handleEditClose={handleEditClose}
        />
      )}
      {deleteModalIsOpen && (
        <ConfirmModal
          affirmativeAction={() => {
            deleteProperty(property);
            setDeleteModalIsOpen(false);
          }}
          negativeText="NO"
          affirmativeText="YES"
          title={`DELETE ${property.label.toUpperCase()}?`}
          onCloseModal={() => {
            setDeleteModalIsOpen(false);
          }}
          isOpen={deleteModalIsOpen}
          negativeAction={() => {
            setDeleteModalIsOpen(false);
          }}
        >
          <div>It will be gone. Forever.</div>
        </ConfirmModal>
      )}
    </>
  );
};

export default OverflowMenuComponent;
