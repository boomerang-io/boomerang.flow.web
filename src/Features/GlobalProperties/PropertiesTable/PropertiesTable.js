import React, { useState } from "react";
import PropTypes from "prop-types";
import matchSorter from "match-sorter";
import { useMutation, queryCache } from "react-query";
import { DataTable, Search, Pagination } from "@boomerang-io/carbon-addons-boomerang-react";
import { Error404, notify, ToastNotification } from "@boomerang-io/carbon-addons-boomerang-react";
import ActionsMenu from "./ActionsMenu";
import CreateEditPropertiesModal from "./CreateEditPropertiesModal";
import Header from "Components/Header";
import { arrayPagination } from "Utils/arrayHelper";
import { stringToPassword } from "Utils/stringHelper";
import { InputType } from "Constants";
import { formatErrorMessage } from "@boomerang-io/utils";
import { serviceUrl, resolver } from "Config/servicesConfig";
import { Checkmark32, Close32 } from "@carbon/icons-react";
import styles from "./propertiesTable.module.scss";

const DEFAULT_PAGE_SIZE = 10;
const PAGE_SIZES = [DEFAULT_PAGE_SIZE, 25, 50];

const configUrl = serviceUrl.getGlobalConfiguration();

function PropertiesTable({ properties }) {
  const [page, setPage] = useState(1);
  const [pageSize, setPageSize] = useState(DEFAULT_PAGE_SIZE);
  const [searchQuery, setSearchQuery] = useState("");
  const [sort, setSort] = useState({ key: "label", sortDirection: "ASC" });

  /** Delete Property */
  const [deleteGlobalPropertyMutation] = useMutation(resolver.deleteGlobalPropertyRequest, {
    onSuccess: () => queryCache.invalidateQueries(configUrl),
  });

  const headers = [
    {
      header: "Label",
      key: "label",
    },
    {
      header: "Key",
      key: "key",
    },
    {
      header: "Description",
      key: "description",
    },
    {
      header: "Value",
      key: "value",
    },
    {
      header: "Secured",
      key: "secured",
    },
    {
      header: "",
      key: "actions",
    },
  ];

  const deleteProperty = async (property) => {
    try {
      await deleteGlobalPropertyMutation({ id: property.id });
      notify(
        <ToastNotification
          kind="success"
          title={"Property Deleted"}
          subtitle={`Request to delete ${property.label} succeeded`}
          data-testid="delete-prop-notification"
        />
      );
    } catch (err) {
      const errorMessages = formatErrorMessage({
        error: err,
        defaultMessage: "Delete Property Failed",
      });
      notify(
        <ToastNotification
          kind="error"
          title={errorMessages.title}
          subtitle={errorMessages.message}
          data-testid="delete-prop-notification"
        />
      );
    }
  };

  const handleSearchChange = (e) => {
    const searchQuery = e.target.value;
    setSearchQuery(searchQuery);
  };

  const handlePaginationChange = ({ page, pageSize }) => {
    setPage(page);
    setPageSize(pageSize);
  };

  const renderCell = (propertyId, cellIndex, value) => {
    const property = properties.find((property) => property.id === propertyId);
    const column = headers[cellIndex];

    switch (column.key) {
      case "value":
        const determineValue = value
          ? property && property.type === InputType.Password
            ? stringToPassword(value)
            : value
          : "---";
        return <p className={styles.tableTextarea}>{determineValue}</p>;
      case "secured":
        return property && property.type === InputType.Password ? (
          <Checkmark32 alt="secured" className={`${styles.tableSecured} ${styles.secured}`} />
        ) : (
            <Close32 alt="unsecured" className={`${styles.tableSecured} ${styles.unsecured}`} />
          );
      case "actions":
        return <ActionsMenu deleteProperty={deleteProperty} property={property} properties={properties} />;
      default:
        return <p className={styles.tableTextarea}>{value || "---"}</p>;
    }
  };

  const handleSort = (valueA, valueB, config) => {
    setSort(config);
  };

  const newProperties = searchQuery
    ? matchSorter(properties, searchQuery, { keys: ["label", "key", "description"] })
    : properties;

  const { TableContainer, Table, TableHead, TableRow, TableBody, TableCell, TableHeader } = DataTable;

  const totalItems = newProperties.length;

  return (
    <>
      <Header
        includeBorder
        title="Properties"
        description="Set global properties that are available for all Workflows."
      ></Header>
      <div className={styles.tableContainer}>
        <div className={styles.header}>
          <Search
            className={styles.search}
            id="properties-table-search"
            labelText="Search"
            placeHolderText="Search"
            onChange={handleSearchChange}
          />
          <div className={styles.actions}>
            <CreateEditPropertiesModal properties={properties} />
          </div>
        </div>
        {totalItems > 0 ? (
          <>
            <DataTable
              rows={arrayPagination(newProperties, page, pageSize, sort)}
              sortRow={handleSort}
              headers={headers}
              render={({ rows, headers, getHeaderProps }) => (
                <TableContainer>
                  <Table>
                    <TableHead>
                      <TableRow className={styles.tableHeadRow}>
                        {headers.map((header) => (
                          <TableHeader
                            id={header.key}
                            {...getHeaderProps({
                              header,
                              className: `${styles.tableHeadHeader} ${styles[header.key]}`,
                            })}
                          >
                            {header.header}
                          </TableHeader>
                        ))}
                      </TableRow>
                    </TableHead>
                    <TableBody className={styles.tableBody}>
                      {rows.map((row) => (
                        <TableRow key={row.id} data-testid="configuration-property-table-row">
                          {row.cells.map((cell, cellIndex) => (
                            <TableCell key={cell.id} style={{ padding: "0" }}>
                              <div className={styles.tableCell}>{renderCell(row.id, cellIndex, cell.value)}</div>
                            </TableCell>
                          ))}
                        </TableRow>
                      ))}
                    </TableBody>
                  </Table>
                </TableContainer>
              )}
            />
            <Pagination
              onChange={handlePaginationChange}
              page={page}
              pageSize={pageSize}
              pageSizes={PAGE_SIZES}
              totalItems={totalItems}
            />
          </>
        ) : (
            <>
              <DataTable
                rows={newProperties}
                headers={headers}
                render={({ headers }) => (
                  <TableContainer>
                    <Table>
                      <TableHead>
                        <TableRow className={styles.tableHeadRow}>
                          {headers.map((header) => (
                            <TableHeader
                              key={header.key}
                              id={header.key}
                              className={`${styles.tableHeadHeader} ${styles[header.key]}`}
                            >
                              {header.header}
                            </TableHeader>
                          ))}
                        </TableRow>
                      </TableHead>
                    </Table>
                  </TableContainer>
                )}
              />
              <Error404 header={null} title="No properties to be found" message={null} />
            </>
          )}
      </div>
    </>
  );
}

PropertiesTable.propTypes = {
  properties: PropTypes.array,
};

export default PropertiesTable;
