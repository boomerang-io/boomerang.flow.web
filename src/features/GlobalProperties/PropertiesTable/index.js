import React, { Component, Fragment } from "react";
import axios from "axios";
import PropTypes from "prop-types";
import matchSorter from "match-sorter";
import { DataTable, Search, Pagination } from "carbon-components-react";
import { notify, ToastNotification, NoDisplay } from "@boomerang/carbon-addons-boomerang-react";
import CreateEditPropertiesModal from "./CreateEditPropertiesModal";
import ActionsMenu from "./ActionsMenu";
import Header from "Components/Header";
import CheckIcon from "@carbon/icons-react/lib/checkmark/32";
import CloseIcon from "@carbon/icons-react/lib/close/32";
import { arrayPagination } from "Utilities/arrayHelper";
import { stringToPassword } from "Utilities/stringHelper";
import INPUT_TYPES from "Constants/inputTypes";
import { BASE_SERVICE_URL } from "Config/servicesConfig";
import styles from "./propertiesTable.module.scss";

const DEFAULT_PAGE_SIZE = 10;
const PAGE_SIZES = [DEFAULT_PAGE_SIZE, 25, 50];

class PropertiesTable extends Component {
  static propTypes = {
    addPropertyInStore: PropTypes.func.isRequired,
    deletePropertyInStore: PropTypes.func.isRequired,
    properties: PropTypes.array,
    updatePropertyInStore: PropTypes.func.isRequired
  };

  state = {
    page: 1,
    pageSize: DEFAULT_PAGE_SIZE,
    properties: this.props.properties,
    sort: {
      key: "label",
      sortDirection: "ASC"
    }
  };

  headers = [
    {
      header: "Label",
      key: "label"
    },
    {
      header: "Key",
      key: "key"
    },
    {
      header: "Description",
      key: "description"
    },
    {
      header: "Value",
      key: "value"
    },
    {
      header: "Secured",
      key: "secured"
    },
    {
      header: "Actions",
      key: "actions "
    }
  ];

  /* Standard table configuration after search or service call */
  resetTableWithNewProperties = properties => {
    const { page, pageSize } = this.state;
    const newPage = page !== 1 && properties.length < pageSize * (page - 1) + 1 ? page - 1 : page;
    this.setState({ page: newPage, properties });
  };

  addPropertyInStore = property => {
    this.props.addPropertyInStore(property);
    const newProperties = [...this.state.properties, property];
    this.resetTableWithNewProperties(newProperties);
  };

  deletePropertyInStore = propertyId => {
    this.props.deletePropertyInStore(propertyId);
    const properties = [...this.state.properties];
    const newProperties = properties.filter(property => property.id !== propertyId);
    this.resetTableWithNewProperties(newProperties);
  };

  updatePropertyInStore = updatedProperty => {
    this.props.updatePropertyInStore(updatedProperty);
    const properties = [...this.state.properties];
    const newProperties = properties.map(property => (property.id === updatedProperty.id ? updatedProperty : property));
    this.resetTableWithNewProperties(newProperties);
  };

  deleteProperty = property => {
    axios
      .delete(`${BASE_SERVICE_URL}/config/${property.id}`)
      .then(response => {
        this.deletePropertyInStore(property.id);
        notify(
          <ToastNotification
            kind="success"
            title={"Property Deleted"}
            subtitle={`Request to delete ${property.label} succeeded`}
          />
        );
      })
      .catch(error => {
        notify(<ToastNotification kind="error" title={"Delete Property Failed"} subtitle={"Something went wrong"} />);
      });
  };

  handleSearchChange = e => {
    const searchQuery = e.target.value;
    const { properties } = this.props;

    const newProperties = searchQuery
      ? matchSorter(properties, searchQuery, { keys: ["label", "key", "description"] })
      : properties;

    this.resetTableWithNewProperties(newProperties);
  };

  handlePaginationChange = ({ page, pageSize }) => {
    this.setState({ page, pageSize });
  };

  renderCell = (propertyId, cellIndex, value) => {
    const property = this.props.properties.find(property => property.id === propertyId);
    const column = this.headers[cellIndex];

    switch (column.header) {
      case "Value":
        const determineValue = value
          ? property && property.type === INPUT_TYPES.PASSWORD
            ? stringToPassword(value)
            : value
          : "---";
        return <p className={styles.tableTextarea}>{determineValue}</p>;
      case "Secured":
        return property && property.type === INPUT_TYPES.PASSWORD ? (
          <CheckIcon alt="secured" className={`${styles.tableSecured} ${styles.secured}`} />
        ) : (
          <CloseIcon alt="unsecured" className={`${styles.tableSecured} ${styles.unsecured}`} />
        );
      case "Actions":
        return (
          <ActionsMenu
            flipped
            addPropertyInStore={this.addPropertyInStore}
            deleteProperty={this.deleteProperty}
            property={property}
            properties={this.props.properties}
            updatePropertyInStore={this.updatePropertyInStore}
          />
        );
      default:
        return <p className={styles.tableTextarea}>{value || "---"}</p>;
    }
  };

  handleSort = (valueA, valueB, config) => {
    this.setState({ sort: config });
  };

  render() {
    const { page, pageSize, properties, sort } = this.state;
    const { TableContainer, Table, TableHead, TableRow, TableBody, TableCell, TableHeader } = DataTable;

    const totalItems = properties.length;

    return (
      <>
        <Header title="Properties" description="Set global properties that are available for all Workflows" />
        <div className={styles.tableContainer}>
          <div className={styles.header}>
            <Search
              className={styles.search}
              id="properties-table-search"
              labelText="Search"
              placeHolderText="Search"
              onChange={this.handleSearchChange}
            />
            <CreateEditPropertiesModal
              addPropertyInStore={this.addPropertyInStore}
              properties={this.props.properties}
              updatePropertyInStore={this.updatePropertyInStore}
            />
          </div>
          {totalItems > 0 ? (
            <>
              <DataTable
                rows={arrayPagination(properties, page, pageSize, sort)}
                sortRow={this.handleSort}
                headers={this.headers}
                render={({ rows, headers, getHeaderProps }) => (
                  <TableContainer>
                    <Table>
                      <TableHead>
                        <TableRow>
                          {headers.map(header => (
                            <TableHeader
                              id={header.key}
                              key={header.key}
                              {...getHeaderProps({
                                header,
                                className: `${styles.tableHeadHeader} ${styles[header.key]}`
                              })}
                            >
                              {header.header}
                            </TableHeader>
                          ))}
                        </TableRow>
                      </TableHead>
                      <TableBody className={styles.tableBody}>
                        {rows.map(row => (
                          <TableRow key={row.id} data-testid="configuration-property-table-row">
                            {row.cells.map((cell, cellIndex) => (
                              <TableCell key={cell.id} style={{ padding: "0" }}>
                                <div className={styles.tableCell}>{this.renderCell(row.id, cellIndex, cell.value)}</div>
                              </TableCell>
                            ))}
                          </TableRow>
                        ))}
                      </TableBody>
                    </Table>
                  </TableContainer>
                )}
              />
              <Pagination
                onChange={this.handlePaginationChange}
                page={page}
                pageSize={pageSize}
                pageSizes={PAGE_SIZES}
                totalItems={totalItems}
              />
            </>
          ) : (
            <Fragment>
              <DataTable
                rows={properties}
                headers={this.headers}
                render={({ headers }) => (
                  <TableContainer>
                    <Table>
                      <TableHead>
                        <TableRow>
                          {headers.map(header => (
                            <TableHeader id={header.key} key={header.key}>{header.header}</TableHeader>
                          ))}
                        </TableRow>
                      </TableHead>
                    </Table>
                  </TableContainer>
                )}
              />
              <NoDisplay
                textLocation="below"
                text={
                  this.props.properties.length > 0
                    ? "No properties found"
                    : "Looks like there aren't any properties. Create one above!"
                }
                style={{ marginTop: "5rem", height: "30rem" }}
              />
            </Fragment>
          )}
        </div>
      </>
    );
  }
}

export default PropertiesTable;
