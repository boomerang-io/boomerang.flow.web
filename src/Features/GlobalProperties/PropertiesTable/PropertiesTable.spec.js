import React from "react";
import PropertiesTable from "./index";
// import config from "../../../apiServer/fixtures/config.js"
import config from "ApiServer/fixtures/config.js";
import { queryCaches } from "react-query";

afterEach(() => {
  queryCaches.forEach((queryCache) => queryCache.clear());
});

describe("PropertiesTable --- Snapshot Test", () => {
  it("Capturing Snapshot of PropertiesTable", () => {
    const { baseElement } = rtlRender(<PropertiesTable properties={config} />);
    expect(baseElement).toMatchSnapshot();
  });
});
