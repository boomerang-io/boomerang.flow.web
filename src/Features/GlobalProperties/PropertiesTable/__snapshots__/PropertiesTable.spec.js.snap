// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertiesTable --- Snapshot Test Capturing Snapshot of PropertiesTable 1`] = `
<body
  id="app"
>
  <div>
    <div
      class="container"
    >
      <div
        class="content"
      >
        <div
          class="info"
        >
          <h1
            class="title"
          >
            Properties
          </h1>
          <p
            class="description"
          >
            Set global properties that are available for all Workflows.
          </p>
        </div>
        <section
          class="buttons"
        />
      </div>
    </div>
    <div
      class="tableContainer"
    >
      <div
        class="header"
      >
        <div
          aria-labelledby="properties-table-search-search"
          class="bx--search bx--search--xl search"
          role="search"
        >
          <svg
            aria-hidden="true"
            class="bx--search-magnifier"
            fill="currentColor"
            focusable="false"
            height="16"
            preserveAspectRatio="xMidYMid meet"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15,14.3L10.7,10c1.9-2.3,1.6-5.8-0.7-7.7S4.2,0.7,2.3,3S0.7,8.8,3,10.7c2,1.7,5,1.7,7,0l4.3,4.3L15,14.3z M2,6.5	C2,4,4,2,6.5,2S11,4,11,6.5S9,11,6.5,11S2,9,2,6.5z"
            />
          </svg>
          <label
            class="bx--label"
            for="properties-table-search"
            id="properties-table-search-search"
          >
            Search
          </label>
          <input
            autocomplete="off"
            class="bx--search-input"
            id="properties-table-search"
            placeholder="Search"
            role="searchbox"
            type="text"
          />
          <button
            aria-label="Clear search input"
            class="bx--search-close bx--search-close--hidden"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              height="20"
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 32 32"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4 14.6 16 8 22.6 9.4 24 16 17.4 22.6 24 24 22.6 17.4 16 24 9.4z"
              />
            </svg>
          </button>
        </div>
        <div
          class="actions"
        >
          <button
            class="bx--btn bx--btn--field bx--btn--primary"
            data-testid="create-global-property-button"
            style="width: 12rem;"
            tabindex="0"
            type="button"
          >
            Create Property
            <svg
              aria-hidden="true"
              aria-label="Create Property"
              class="bx--btn__icon"
              fill="currentColor"
              focusable="false"
              height="16"
              preserveAspectRatio="xMidYMid meet"
              role="img"
              viewBox="0 0 32 32"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M17 15L17 8 15 8 15 15 8 15 8 17 15 17 15 24 17 24 17 17 24 17 24 15z"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="bx--data-table-container"
      >
        <table
          class="bx--data-table bx--data-table--no-border"
        >
          <thead>
            <tr
              class="tableHeadRow"
            >
              <th
                class="tableHeadHeader label"
                id="label"
                scope="col"
              >
                <span
                  class="bx--table-header-label"
                >
                  Label
                </span>
              </th>
              <th
                class="tableHeadHeader key"
                id="key"
                scope="col"
              >
                <span
                  class="bx--table-header-label"
                >
                  Key
                </span>
              </th>
              <th
                class="tableHeadHeader description"
                id="description"
                scope="col"
              >
                <span
                  class="bx--table-header-label"
                >
                  Description
                </span>
              </th>
              <th
                class="tableHeadHeader value"
                id="value"
                scope="col"
              >
                <span
                  class="bx--table-header-label"
                >
                  Value
                </span>
              </th>
              <th
                class="tableHeadHeader secured"
                id="secured"
                scope="col"
              >
                <span
                  class="bx--table-header-label"
                >
                  Secured
                </span>
              </th>
              <th
                class="tableHeadHeader actions"
                id="actions"
                scope="col"
              >
                <span
                  class="bx--table-header-label"
                />
              </th>
            </tr>
          </thead>
          <tbody
            aria-live="polite"
            class="tableBody"
          >
            <tr
              data-testid="configuration-property-table-row"
            >
              <td
                style="padding: 0px;"
              >
                <div
                  class="tableCell"
                >
                  <p
                    class="tableTextarea"
                  >
                    Test property
                  </p>
                </div>
              </td>
              <td
                style="padding: 0px;"
              >
                <div
                  class="tableCell"
                >
                  <p
                    class="tableTextarea"
                  >
                    test.key
                  </p>
                </div>
              </td>
              <td
                style="padding: 0px;"
              >
                <div
                  class="tableCell"
                >
                  <p
                    class="tableTextarea"
                  >
                    adding a test description
                  </p>
                </div>
              </td>
              <td
                style="padding: 0px;"
              >
                <div
                  class="tableCell"
                >
                  <p
                    class="tableTextarea"
                  >
                    Test value
                  </p>
                </div>
              </td>
              <td
                style="padding: 0px;"
              >
                <div
                  class="tableCell"
                >
                  <svg
                    alt="unsecured"
                    aria-hidden="true"
                    class="tableSecured unsecured"
                    fill="currentColor"
                    focusable="false"
                    height="32"
                    preserveAspectRatio="xMidYMid meet"
                    viewBox="0 0 32 32"
                    width="32"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4 14.6 16 8 22.6 9.4 24 16 17.4 22.6 24 24 22.6 17.4 16 24 9.4z"
                    />
                  </svg>
                </div>
              </td>
              <td
                style="padding: 0px;"
              >
                <div
                  class="tableCell"
                >
                  <button
                    aria-expanded="false"
                    aria-haspopup="true"
                    aria-label="Overflow menu"
                    class="bx--overflow-menu"
                    data-testid="configuration-property-table-overflow-menu"
                    tabindex="0"
                  >
                    <svg
                      aria-label="Overflow menu icon"
                      class="bx--overflow-menu__icon"
                      fill="currentColor"
                      focusable="false"
                      height="16"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      viewBox="0 0 32 32"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <circle
                        cx="16"
                        cy="8"
                        r="2"
                      />
                      <circle
                        cx="16"
                        cy="16"
                        r="2"
                      />
                      <circle
                        cx="16"
                        cy="24"
                        r="2"
                      />
                      <title>
                        Overflow menu icon
                      </title>
                    </svg>
                  </button>
                </div>
              </td>
            </tr>
            <tr
              data-testid="configuration-property-table-row"
            >
              <td
                style="padding: 0px;"
              >
                <div
                  class="tableCell"
                >
                  <p
                    class="tableTextarea"
                  >
                    ben.label
                  </p>
                </div>
              </td>
              <td
                style="padding: 0px;"
              >
                <div
                  class="tableCell"
                >
                  <p
                    class="tableTextarea"
                  >
                    ben.key
                  </p>
                </div>
              </td>
              <td
                style="padding: 0px;"
              >
                <div
                  class="tableCell"
                >
                  <p
                    class="tableTextarea"
                  >
                    ---
                  </p>
                </div>
              </td>
              <td
                style="padding: 0px;"
              >
                <div
                  class="tableCell"
                >
                  <p
                    class="tableTextarea"
                  >
                    •••••••••
                  </p>
                </div>
              </td>
              <td
                style="padding: 0px;"
              >
                <div
                  class="tableCell"
                >
                  <svg
                    alt="secured"
                    aria-hidden="true"
                    class="tableSecured secured"
                    fill="currentColor"
                    focusable="false"
                    height="32"
                    preserveAspectRatio="xMidYMid meet"
                    viewBox="0 0 32 32"
                    width="32"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M13 21.2L5.4 13.6 4 15 13 24 28 9 26.6 7.5 13 21.2z"
                    />
                  </svg>
                </div>
              </td>
              <td
                style="padding: 0px;"
              >
                <div
                  class="tableCell"
                >
                  <button
                    aria-expanded="false"
                    aria-haspopup="true"
                    aria-label="Overflow menu"
                    class="bx--overflow-menu"
                    data-testid="configuration-property-table-overflow-menu"
                    tabindex="0"
                  >
                    <svg
                      aria-label="Overflow menu icon"
                      class="bx--overflow-menu__icon"
                      fill="currentColor"
                      focusable="false"
                      height="16"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      viewBox="0 0 32 32"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <circle
                        cx="16"
                        cy="8"
                        r="2"
                      />
                      <circle
                        cx="16"
                        cy="16"
                        r="2"
                      />
                      <circle
                        cx="16"
                        cy="24"
                        r="2"
                      />
                      <title>
                        Overflow menu icon
                      </title>
                    </svg>
                  </button>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div
        class="bx--pagination"
      >
        <div
          class="bx--pagination__left"
        >
          <label
            class="bx--pagination__text"
            for="bx-pagination-select-1"
            id="bx-pagination-select-1-count-label"
          >
            Items per page:
          </label>
          <div
            class="bx--form-item"
          >
            <div
              class="bx--select bx--select--inline bx--select__item-count"
            >
              <div
                class="bx--select-input--inline__wrapper"
              >
                <div
                  class="bx--select-input__wrapper"
                >
                  <select
                    class="bx--select-input"
                    id="bx-pagination-select-1"
                  >
                    <option
                      class="bx--select-option"
                      value="10"
                    >
                      10
                    </option>
                    <option
                      class="bx--select-option"
                      value="25"
                    >
                      25
                    </option>
                    <option
                      class="bx--select-option"
                      value="50"
                    >
                      50
                    </option>
                  </select>
                  <svg
                    aria-hidden="true"
                    class="bx--select__arrow"
                    fill="currentColor"
                    focusable="false"
                    height="16"
                    preserveAspectRatio="xMidYMid meet"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
          <span
            class="bx--pagination__text"
          >
            1–2 of 2 items
          </span>
        </div>
        <div
          class="bx--pagination__right"
        >
          <div
            class="bx--form-item"
          >
            <div
              class="bx--select bx--select--inline bx--select__page-number"
            >
              <label
                class="bx--label bx--visually-hidden"
                for="bx-pagination-select-1-right"
              >
                Page number, of 1 pages
              </label>
              <div
                class="bx--select-input--inline__wrapper"
              >
                <div
                  class="bx--select-input__wrapper"
                >
                  <select
                    class="bx--select-input"
                    id="bx-pagination-select-1-right"
                  >
                    <option
                      class="bx--select-option"
                      value="1"
                    >
                      1
                    </option>
                  </select>
                  <svg
                    aria-hidden="true"
                    class="bx--select__arrow"
                    fill="currentColor"
                    focusable="false"
                    height="16"
                    preserveAspectRatio="xMidYMid meet"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
          <span
            class="bx--pagination__text"
          >
            of 1 pages
          </span>
          <button
            aria-label="Previous page"
            class="bx--pagination__button bx--pagination__button--backward bx--pagination__button--no-index"
            disabled=""
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              height="24"
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 32 32"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M20 24L10 16 20 8z"
              />
            </svg>
          </button>
          <button
            aria-label="Next page"
            class="bx--pagination__button bx--pagination__button--forward bx--pagination__button--no-index"
            disabled=""
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              height="24"
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 32 32"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 8L22 16 12 24z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="bx--bmrg-modal-portal"
  />
  <div
    class="bx--bmrg-modal-portal"
  />
  <div
    class="bx--bmrg-modal-portal"
  />
  <div
    class="bx--bmrg-modal-portal"
  />
  <div
    class="bx--bmrg-modal-portal"
  />
</body>
`;
